server/lib/src/idm/oauth2.rs

```diff
@@ -1819,67 +1830,57 @@ impl<'a> IdmServerProxyReadTransaction<'a> {
                 Oauth2Error::InvalidClientId
             })?;
 
-        let allow_localhost_redirect = match &o2rs.type_ {
-            OauthRSType::Basic { .. } => false,
-            OauthRSType::Public {
-                allow_localhost_redirect,
-            } => *allow_localhost_redirect,
-        };
+        // redirect_uri must be part of the client_id origins, unless the client is public and then it MAY
+        // be a loopback address exempting it from this check and enforcement.
+        let is_loopback_redirect = o2rs.type_.allow_localhost_redirect()
+            && check_loopback_matches(&auth_req.redirect_uri, &o2rs.redirect_uris);
+
+        // if they're doing localhost things, then we can carry
+        if !is_loopback_redirect {
+            // The legacy origin match is in use.
+            let origin_uri_matched =
+                !o2rs.strict_redirect_uri && o2rs.origins.contains(&auth_req.redirect_uri.origin());
+
+            // Strict uri validation is in use.
+            let strict_redirect_uri_matched =
+                o2rs.strict_redirect_uri && o2rs.redirect_uris.contains(&auth_req.redirect_uri);
+            // Allow opaque origins such as app uris.
+            let opaque_origin_matched = o2rs.opaque_origins.contains(&auth_req.redirect_uri);
+
+            // At least one of these conditions must hold true to proceed.
```

Comment:
I don't like this re-arrangement, it makes it harder to follow the logic. Before it was split up to check what conditions might be true, then we assert the combination we need.

Reply:
I'll try another run at it - it got pretty convoluted hence not doing the checks until they were needed.